### ISSUE TITLE
[$500] Chat - Link in end of line displays tooltip over text and not on link

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#27585]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the app
2. Open any report
3. Type in some text and at the end of line, add link with long text eg: [Test test test test test test test test test test]([testing.com](http://testing.com/))
4. Send the text, text should have text and link in first line and link should be continued in second line too
5. Hover on link and observe that tooltip is displayed over text

## Expected Result:
App should display tooltip over link on hover

## Actual Result:
 App displays tooltip over text on link hover when link is multiline and occupies the end of line with text in start

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.70-5

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/93edea85-b4a3-4757-a14b-8d2ca5676956

https://github.com/Expensify/App/assets/78819774/8aae4252-8177-4a90-8a3d-ae482334dd10

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Dhanashree-Sawant

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1694604089052599

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01aa77f02fd5409216</li>
        <li>Upwork Job ID: 1703100827931217920</li>
        <li>Last Price Increase: 2023-09-16</li>
<li>Automatic offers: </li>
<ul>
<li>s77rt | Reviewer | 26768759</li>
<li>Dhanashree-Sawant | Reporter | 26768761</li>
</ul></li>
    </ul>
</details>

**Comment by ewanmellor** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When a link is wrapped onto multiple lines in a chat window, and the user hovers over the link, the tooltip appears in the wrong place.

### What is the root cause of that problem?

The problem is that the `Tooltip` component places the tooltip over the bounding box of the element that is being hovered. For a link that has wrapped onto two lines, the bounding box encloses the entire two lines, and so it looks like the tooltip is being placed over the paragraph of text instead of the link.

### What changes do you think we should make in order to solve the problem?

From the user's perspective, you want the tooltip to target the link, but since the link is now in two parts (one at the end of the first line, one at the start of the second line) you want the tooltip to target the center of whichever part of the link the user is hovering over.

It is possible to get the browser to give you the bounding box of each part of the text. However, `Tooltip` uses `@react-ng/bounds-observer` to get the element bounding box, and doesn't do anything to position the tooltip when the user hovers. We would need to extend `Tooltip` to reposition the tooltip in the `onHoverIn` handler, depending on which part of the link the user is hovering over.

This algorithm will add a bit of cost to the `onHoverIn` event, so I'd add a prop to `Tooltip` to make it opt-in.

**Comment by s77rt** [comment]
@ewanmellor Thanks for the proposal. Your RCA is correct. In our App we seem to always position the tooltip on the center of the target regardless of the mouse position. I noticed that Slack do things a little different, they position the tooltip on the center of the _partial_ target, the partial target being the part of the target that the mouse is at. Your suggested solution seems to help towards that approach but it's not much clear. Can you please elaborate your proposal on how the algorithm would work?

**Comment by ewanmellor** [comment]
@s77rt Yes, that's what I meant.

1. Use `Element. getClientRects` to get the bounding boxes for each partial target.
2. Use the mouse position from the `onHoverIn` event to determine which partial target the user is hovering over.
3. Set the tooltip position to center over that partial target.
4. Fall back to the current algorithm if necessary, though `getClientRects` is well supported so it should not be a problem.

**Comment by s77rt** [comment]
@ewanmellor Thanks for the update. That makes sense, can you please update your proposal to include that?

BTW, do you have a proof of concept or did you manage to get this working? not really necessary, just it would help avoid deadends

**Comment by ewanmellor** [comment]
@s77rt I haven't written the code yet, no. I am waiting for approval before I spend the time.

**Comment by ewanmellor** [comment]
@s77rt Proposal updated: https://github.com/Expensify/App/issues/27585#issuecomment-1724380334

**Comment by s77rt** [comment]
@ewanmellor Thanks for the update!

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/27585#issuecomment-1724380334)

**Comment by pecanoro** [comment]
Assigning @ewanmellor, let's see if we can fix this one!

**Comment by pecanoro** [comment]
Reopening the job to contributors since @ewanmellor does not seem to be able to handle this one anymore.

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by sophiepintoraetz** [comment]
Hmm, might drop this in an agency room tomorrow to see if there are any takers.

**Comment by s77rt** [comment]
Not overdue. Looking for proposals

**Comment by sophiepintoraetz** [comment]
[Dropped into Callstack!](https://expensify.slack.com/archives/C03UK30EA1Z/p1696480310660259)

**Comment by sophiepintoraetz** [comment]
Hey @kadiealexander, just need some help babysitting this issue while I'm OOO till 15 Oct:
- We had a contributor assigned and then they silently ghosted the issue, so we're looking for "proposals" but really, anyone could implement the [proposal that was approved](https://github.com/Expensify/App/issues/27585#issuecomment-1726284874).
- I've [asked Callstack](https://expensify.slack.com/archives/C03UK30EA1Z/p1696480310660259) to see if they have capacity 🤞 - in which case, only @s77rt would need payment, as I have already paid the bug reporter @dhanashree-sawant.
- If they don't have capacity, I was going to try Software mansion next!

**Comment by teneeto** [comment]
Hi, I'm Eto from Callstack - expert contributor group - and I would like to take a look at this issue.

**Comment by s77rt** [comment]
Not overdue. We are working on it Melvin

**Comment by ewanmellor** [comment]
> * We had a contributor assigned and then they silently ghosted the issue, so we're looking for "proposals" but really, anyone could implement the [proposal that was approved](https://github.com/Expensify/App/issues/27585#issuecomment-1726284874).

@sophiepintoraetz I didn't "ghost", I withdrew. You are welcome to merge my PR; any of the three versions of it that I submitted, in fact. But if project members here think that spending 3 weeks (and counting) on a change like this is valuable, well I have better things to do with my time.

**Comment by teneeto** [comment]
Please Let's continue Review of the Solution to this Issue with new Information.
------
- New PR: https://github.com/Expensify/App/pull/29134
- Leading PR: https://github.com/Expensify/App/pull/22010

Note: This solution was originally authored by @ewanmellor and for some unknown reasons Review fixes are not complete and PR is closed. It would be nice to assign PR to same reviewer, then continue discussions on the new PR.

**Comment by s77rt** [comment]
Note: We had a regression https://github.com/Expensify/App/issues/29446.

**Comment by s77rt** [comment]
The PR fixing this is merged deployed to production. Melvin forgot to remove the `Reviewing` label here

**Comment by s77rt** [comment]
And another one 😅 https://github.com/Expensify/App/issues/29678

**Comment by pecanoro** [comment]
@s77rt Do you know why it created an additional issue for payment instead of updating this one? Probably the issue wasn't linked properly in the PR description.

**Comment by sophiepintoraetz** [comment]
I've already paid @dhanashree-sawant for this issue, so if that's all that we're waiting on, we can close this issue?

**Comment by pecanoro** [comment]
Yes! We can close it then, thank you!
